lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp

```diff
@@ -1631,26 +1638,48 @@ bool SymbolFileDWARF::CompleteType(CompilerType &compiler_type) {
     return true;
   }
 
-  DWARFDIE dwarf_die = GetDIE(die_it->getSecond());
-  if (dwarf_die) {
-    // Once we start resolving this type, remove it from the forward
-    // declaration map in case anyone child members or other types require this
-    // type to get resolved. The type will get resolved when all of the calls
-    // to SymbolFileDWARF::ResolveClangOpaqueTypeDefinition are done.
-    GetForwardDeclCompilerTypeToDIE().erase(die_it);
+  DWARFDIE decl_die = GetDIE(die_it->getSecond());
+  // Once we start resolving this type, remove it from the forward
+  // declaration map in case anyone's child members or other types require this
+  // type to get resolved.
+  GetForwardDeclCompilerTypeToDIE().erase(die_it);
+  DWARFDIE def_die = FindDefinitionDIE(decl_die);
+  if (!def_die) {
+    SymbolFileDWARFDebugMap *debug_map_symfile = GetDebugMapSymfile();
+    if (debug_map_symfile) {
+      // We weren't able to find a full declaration in this DWARF, see
+      // if we have a declaration anywhere else...
+      def_die = debug_map_symfile->FindDefinitionDIE(decl_die);
+    }
+  }
+  if (!def_die) {
+    // No definition found. Proceed with the declaration die. We can use it to
+    // create a forward-declared type.
+    def_die = decl_die;
+  }
 
-    Type *type = GetDIEToType().lookup(dwarf_die.GetDIE());
+  Type *type = ResolveType(def_die);
```

Comment:
This still creates the lookup loop where we use the type to find the declaration DIE, use the decl DIE to find the definition DIE, and then use that DIE to find the type again. I really wanted to avoid that, for several reasons:
- it's fragile, in that if this call returns a type different from the one we started, we will end up in an inconsistent state. I think this doesn't crash right now because my change to CompleteTypeFromDWARF has made it more robust to these kinds of inconsistent inputs. However, that wasn't my goal, and I don't think we should be relying on that.
-  it's doing more work than it should. We already *know* which type we're supposed to be completing. Why are we looking it up again?
- it makes the code hard to follow

I was hoping this code would basically boil down to something like (with additional error checks, probably):
```
dwarf_ast->CompleteTypeFromDWARF(FindDefinitionDIE(decl_die), GetDIEToType().lookup(decl_die.GetDIE()), compiler_type);
```

Why do we need to do all of the other stuff? Can we refactor something so that we don't need to do all those other things here?

Reply:
`ResolveType(def_die)` just do [this part](https://github.com/llvm/llvm-project/pull/98361/files/37b6878b9125c314c75053f7d5b0ba520111e9a3#diff-5dd6736e4d6c38623630df16c4494c1a8b099716ee0f05c9af54b4bafb1d864eR1607-R1674) of work and then return the type, which was created from the declaration DIE. It doesn't create a loop or recursion. We need it to do the work when finding existing type in the UniqueDWARFASTTypeMap, which is basically this block: https://github.com/llvm/llvm-project/pull/98361/files/37b6878b9125c314c75053f7d5b0ba520111e9a3#diff-5dd6736e4d6c38623630df16c4494c1a8b099716ee0f05c9af54b4bafb1d864eR1648-R1674. Those code are important to keep some maps(`GetDIEToType()`, `LinkDeclContextToDIE`, `UniqueDWARFASTTypeMap`) updated.

Reply:
> We need it to do the work when finding existing type in the UniqueDWARFASTTypeMap,

Why do we need to find that type? We already know which type we're completing (we get it as an argument to `SymbolFileDWARF::CompleteType`)...

> It doesn't create a loop or recursion. We need it to do the work when finding existing type in the UniqueDWARFASTTypeMap, which is basically this block: https://github.com/llvm/llvm-project/pull/98361/files/37b6878b9125c314c75053f7d5b0ba520111e9a3#diff-5dd6736e4d6c38623630df16c4494c1a8b099716ee0f05c9af54b4bafb1d864eR1648-R1674.

That's only true if `UniqueDWARFASTTypeMap` functions perfectly and returns the same type we started out with. My point is we shouldn't/don't need to rely on that.

> Those code are important to keep some maps(`GetDIEToType()`, `LinkDeclContextToDIE`, `UniqueDWARFASTTypeMap`) updated

Yes, we need to update those maps, but I think we can/should just update those directly, instead of relying on it to be populated as a side-effect of the second ParseTypeFromDWARF call.  Like, we already have the type and the definition die, so why don't we just do a `GetDIEToType()[defn_die] = type`? And for the UniqueDWARFASTTypeMap, we can use the original declaration die to perform the lookup. This will be both faster (we can add a specialized lookup function that assumes that the lookup die matches one of the `udt.m_die` members) and guaranteed to be correct.

Reply:
That makes sense. I created another function to update those maps and use decl_die to lookup in UniqueDWARFASTTypeMap.